pkg/build/api/validation/validation.go

```diff
@@ -149,52 +138,25 @@ func hasProxy(source *buildapi.GitBuildSource) bool {
 	return len(source.HTTPProxy) > 0 || len(source.HTTPSProxy) > 0
 }
 
-func validateSource(input *buildapi.BuildSource) fielderrors.ValidationErrorList {
+func validateSource(input *buildapi.BuildSource, isCustomStrategy bool) fielderrors.ValidationErrorList {
 	allErrs := fielderrors.ValidationErrorList{}
-	switch input.Type {
-	case buildapi.BuildSourceGit:
-		if input.Git == nil {
-			allErrs = append(allErrs, fielderrors.NewFieldRequired("git"))
-		} else {
-			allErrs = append(allErrs, validateGitSource(input.Git).Prefix("git")...)
-		}
-		if input.Dockerfile != nil {
-			allErrs = append(allErrs, validateDockerfile(*input.Dockerfile)...)
-		}
-		if input.Binary != nil {
-			allErrs = append(allErrs, fielderrors.NewFieldInvalid("binary", "", "may not be set when type is Git"))
-		}
-	case buildapi.BuildSourceBinary:
-		if input.Binary == nil {
-			allErrs = append(allErrs, fielderrors.NewFieldRequired("binary"))
-		} else {
-			allErrs = append(allErrs, validateBinarySource(input.Binary).Prefix("binary")...)
-		}
-		if input.Dockerfile != nil {
-			allErrs = append(allErrs, validateDockerfile(*input.Dockerfile)...)
-		}
-		if input.Git != nil {
-			allErrs = append(allErrs, fielderrors.NewFieldInvalid("git", "", "may not be set when type is Binary"))
-		}
-	case buildapi.BuildSourceDockerfile:
-		if input.Dockerfile == nil {
-			allErrs = append(allErrs, fielderrors.NewFieldRequired("dockerfile"))
-		} else {
-			allErrs = append(allErrs, validateDockerfile(*input.Dockerfile)...)
+
+	if input.Git != nil {
+		// git and binary input sources are mutually exclusive
+		if input.Binary != nil && !isCustomStrategy {
+			return append(allErrs, fielderrors.NewFieldInvalid("binary", "", "may not be set when git also set"))
 		}
-		switch {
-		case input.Git != nil && input.Binary != nil:
-			allErrs = append(allErrs, fielderrors.NewFieldInvalid("git", "", "may not be set when binary is also set"))
-			allErrs = append(allErrs, fielderrors.NewFieldInvalid("binary", "", "may not be set when git is also set"))
-		case input.Git != nil:
-			allErrs = append(allErrs, validateGitSource(input.Git).Prefix("git")...)
-		case input.Binary != nil:
-			allErrs = append(allErrs, validateBinarySource(input.Binary).Prefix("binary")...)
+		allErrs = append(allErrs, validateGitSource(input.Git).Prefix("git")...)
+	}
+	if input.Binary != nil {
+		// git and binary input sources are mutually exclusive
+		if input.Git != nil && !isCustomStrategy {
```

Comment:
agree, but it's safer to have it there in case someone reorders this later or other refactoring occurs.